### PR TITLE
Gateway API: support regex path and header match

### LINKED ISSUE #3357: Gateway API: "RegularExpression" HTTPHeaderMatch [custom]
For an HTTPRoute, need to implement the "RegularExpression" `HTTPHeaderMatch` on a `HTTPRouteRule.HTTPRouteMatch`.

**TODO: Determine regex type to implement & document properly**

> Since RegularExpression PathType has custom conformance, implementations can support POSIX, PCRE or any other dialects of regular expressions.

// ref: https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.HTTPHeaderMatch

### LINKED ISSUE #5219: Gateway API: support regex path match [custom conformance]
**Please describe the problem you have**
Support the `RegularExpression` path match type (https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1beta1.PathMatchType)

Custom conformance

### PR DESCRIPTION
**Pull request by fangfpeng**
- support regex path and regex header match for HTTPRoute.
- support regex header match for GRPCRoute.

Fixes #3357
Fixes #5219

### CODE REVIEW

File: internal/dag/gatewayapi_processor.go [hunk 1]
```diff
@@ -1551,10 +1564,19 @@ func gatewayHeaderMatchConditions(matches []gatewayapi_v1beta1.HTTPHeaderMatch)
...
+		if match.Type == nil {
+			matchType = HeaderMatchTypeExact
+		} else {
+			switch *match.Type {
+			case gatewayapi_v1beta1.HeaderMatchExact:
+				matchType = HeaderMatchTypeExact
+			case gatewayapi_v1beta1.HeaderMatchRegularExpression:
+				matchType = HeaderMatchTypeRegex
+			default:
+				return nil, fmt.Errorf("HTTPRoute.Spec.Rules.Matches.Headers: Only Exact type and RegularExpressionmatch type are supported")
```
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:1578]: nit:
```suggestion
				return nil, fmt.Errorf("HTTPRoute.Spec.Rules.Matches.Headers: Only Exact type and RegularExpression match type are supported")
```

File: internal/dag/gatewayapi_processor.go [hunk 2]
```diff
@@ -1417,7 +1426,7 @@ func gatewayGRPCHeaderMatchConditions(matches []gatewayapi_v1alpha2.GRPCHeaderMa
 		seenNames.Insert(upperName)
 
 		headerMatchConditions = append(headerMatchConditions, HeaderMatchCondition{
-			MatchType: HeaderMatchTypeExact,
+			MatchType: matchType,
```
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:1429]: For regex, we should validate that it's a valid regex and return an error if not. We have a `ValidateRegex` function in this package that you can use.

File: internal/dag/gatewayapi_processor.go [hunk 3]
```diff
@@ -1534,14 +1543,18 @@ func gatewayPathMatchCondition(match *gatewayapi_v1beta1.HTTPPathMatch, routeAcc
 			return nil, false
 		}
 
+		if *match.Type == gatewayapi_v1beta1.PathMatchRegularExpression {
+			return &RegexMatchCondition{Regex: path}, true
```
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:1547]: Same comment re: validating the regex

File: internal/dag/gatewayapi_processor.go [hunk 4]
```diff
@@ -1524,7 +1533,7 @@ func gatewayPathMatchCondition(match *gatewayapi_v1beta1.HTTPPathMatch, routeAcc
 		return &PrefixMatchCondition{Prefix: path, PrefixMatchType: PrefixMatchSegment}, true
 	}
 
-	if *match.Type == gatewayapi_v1beta1.PathMatchExact {
+	if *match.Type == gatewayapi_v1beta1.PathMatchExact || *match.Type == gatewayapi_v1beta1.PathMatchRegularExpression {
```
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:1536]: I don't think we need to add regex matches here -- these checks only apply for exact or prefix matches:

> PathPrefix and Exact paths must be syntactically valid:
> 
> - Must begin with the / character
> - Must not contain consecutive / characters (e.g. /foo///, //).

(ref. https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1beta1.PathMatchType)

File: internal/dag/gatewayapi_processor.go [hunk 5]
```diff
@@ -1566,7 +1588,7 @@ func gatewayHeaderMatchConditions(matches []gatewayapi_v1beta1.HTTPHeaderMatch)
 		seenNames.Insert(upperName)
 
 		headerMatchConditions = append(headerMatchConditions, HeaderMatchCondition{
-			MatchType: HeaderMatchTypeExact,
+			MatchType: matchType,
 			Name:      string(match.Name),
 			Value:     match.Value,
```
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:1593]: same comment re: validating the regex

**Review by skriss — CHANGES_REQUESTED** [review]
Thanks for this @fangfpeng! Couple comments

File: internal/dag/gatewayapi_processor.go [hunk 6]
```diff
@@ -1402,10 +1402,22 @@ func gatewayGRPCHeaderMatchConditions(matches []gatewayapi_v1alpha2.GRPCHeaderMa
...
-		// "Exact" is the default if not defined in the object, and
-		// the only supported match type.
-		if match.Type != nil && *match.Type != gatewayapi_v1beta1.HeaderMatchExact {
-			return nil, fmt.Errorf("GRPCRoute.Spec.Rules.Matches.Headers: Only Exact match type is supported")
+		// "Exact" and "RegularExpression" are the only supported match types. If match type is not specified, use "Exact" as default.
+		var matchType string
+		if match.Type == nil {
+			matchType = HeaderMatchTypeExact
+		} else {
+			switch *match.Type {
```
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:1410]: Could simplify this a bit using a generic helper we have for defaulting a nil pointer:
```suggestion
		switch ref.Val(match.Type, gatewayapi_v1beta1.HeaderMatchExact) {
```

File: internal/dag/gatewayapi_processor.go [hunk 7]
```diff
@@ -1551,10 +1571,22 @@ func gatewayHeaderMatchConditions(matches []gatewayapi_v1beta1.HTTPHeaderMatch)
...
-		// "Exact" is the default if not defined in the object, and
-		// the only supported match type.
-		if match.Type != nil && *match.Type != gatewayapi_v1beta1.HeaderMatchExact {
-			return nil, fmt.Errorf("HTTPRoute.Spec.Rules.Matches.Headers: Only Exact match type is supported")
+		// "Exact" and "RegularExpression" are the only supported match types. If match type is not specified, use "Exact" as default.
+		var matchType string
+		if match.Type == nil {
+			matchType = HeaderMatchTypeExact
+		} else {
+			switch *match.Type {
```
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:1579]: Same simplification here:
```suggestion
		switch ref.Val(match.Type, gatewayapi_v1beta1.HeaderMatchExact) {
```

**Review by skriss — CHANGES_REQUESTED** [review]

**Review by skriss — APPROVED** [review]
Thanks for the updates @fangfpeng, LGTM!

File: internal/dag/builder_test.go [hunk 8]
```diff
@@ -14581,6 +14659,14 @@ func exactrouteHTTPRoute(path string, first *Service, rest ...*Service) *Route {
 	}
 }
 
+func regrexrouteHTTPRoute(path string, first *Service, rest ...*Service) *Route {
```
Comment (sunjayBhatia) [inline internal/dag/builder_test.go:14662]: tiny typo nit
```suggestion
func regexrouteHTTPRoute(path string, first *Service, rest ...*Service) *Route {
```

**Pull request merged**